### PR TITLE
Introduce `sort_unique`

### PR DESCRIPTION
**Pull request by schnellerhase**
While working on https://github.com/FEniCS/dolfinx/pull/3293, noticed that a lot of the use cases of `sort` are coupled with a following `unique` call to produce a sorted and unique range.

Introduces `common::sort_unique` which sorts and uniquifies a range, given a custom sort routine. To allow for the passing of `radix_sort` its interface is made `std::ranges` compatible, i.e. functionality is exposed by `opeartor()` of a struct.

### CODE REVIEW

**Comment by jhale** [comment]
I like the ranges port of the radix sort in principle, although not reviewed.

Not sure about wrapping sort/unique/raise - yes it's a bit wordy, but it's canonical STL (we didn't pick C++ for conciseness ;)).

**Comment by garth-wells** [comment]
Likewise, like the look of the radix sort but prefer to not wrap STL functionality.

**Comment by schnellerhase** [comment]
Ok, so I'll remove the `sort_unique`. I'll open another PR with the `std::ranges`` update for the `std::unique`, left this out previously to try this first.

File: cpp/dolfinx/common/sort.h [hunk 1]
```diff
@@ -23,144 +27,81 @@ namespace dolfinx
...
+    if (its % 2 != 0)
+      std::ranges::copy(buffer, range.begin());
   }
+};
 
-  if (its % 2 == 1)
-    std::ranges::copy(perm2, perm.begin());
-}
+/// Radix sort
+inline constexpr __radix_sort radix_sort{};
```
Comment (jhale) [inline cpp/dolfinx/common/sort.h:104]: Is this `inline` necessary?

File: cpp/dolfinx/common/sort.h [hunk 2]
```diff
@@ -23,144 +27,81 @@ namespace dolfinx
...
+    using I = std::iter_value_t<R>;
 
-  if (array.size() <= 1)
-    return;
+    // value type (if no projection is provided it holds I == T)
+    using T = std::remove_cvref_t<std::result_of_t<P(I)>>;
 
-  const auto [min, max] = std::ranges::minmax_element(array);
-  T range = *max - *min + 1;
+    static_assert(std::is_integral_v<T>, "This function only sorts integers.");
```
Comment (jhale) [inline cpp/dolfinx/common/sort.h:42]: Can this be done by applying a concept/restriction to `R` or `P`?

File: cpp/dolfinx/common/sort.h [hunk 3]
```diff
@@ -23,144 +27,81 @@ namespace dolfinx
 /// @tparam T Integral type
```
Comment (jhale) [inline cpp/dolfinx/common/sort.h:27]: Documentation no longer correct.

File: cpp/dolfinx/common/sort.h [hunk 4]
```diff
@@ -23,144 +27,81 @@ namespace dolfinx
...
 /// @tparam BITS The number of bits to sort at a time.
 /// @param[in, out] array The array to sort.
-template <typename T, int BITS = 8>
-void radix_sort(std::span<T> array)
+struct __radix_sort
```
Comment (jhale) [inline cpp/dolfinx/common/sort.h:30]: Is it necessary to pack the method into a struct?

**Comment by jhale** [comment]
Very nice work, a few comments that are possibly more out of ignorance.

File: cpp/dolfinx/common/sort.h (hunk 4 above)
Comment (schnellerhase) [reply cpp/dolfinx/common/sort.h:30]: This follows the style of the STL regarding the handling of functors and enables to use the `radix_sort` in any other STL algorithm as an inpute function.

The new design with theses structs follows the realization of lambdas, and removes the necessity of handling function pointers as template types. This reduces the amount of necessary type dispatching drastically if I remember correctly.

What we could do is move the implementation of the `radix_sort` into an anonymous namespace to separate it from its wrapping for external use.

File: cpp/dolfinx/common/sort.h (hunk 1 above)
Comment (schnellerhase) [reply cpp/dolfinx/common/sort.h:104]: I believe so - every time we include the `sort.h` we will have a different definition of `radix_sort` so we must allow for multiple definitions.

File: cpp/dolfinx/common/sort.h (hunk 4 above)
Comment (jhale) [reply cpp/dolfinx/common/sort.h:24]: Thanks for educating me!

File: cpp/dolfinx/common/sort.h [hunk 5]
```diff
@@ -6,161 +6,105 @@
...
-      counter[bucket]++;
+      max_value >>= BITS;
+      its++;
     }
 
-    // Prefix sum to get the inserting position
-    offset[0] = 0;
-    std::partial_sum(counter.begin(), counter.end(), std::next(offset.begin()));
+    // Adjacency list arrays for computing insertion position
+    std::array<std::int32_t, bucket_size> counter;
```
Comment (jhale) [inline cpp/dolfinx/common/sort.h:66]: Are these`std::int32_t` always OK? Seems strange for a generic algorithm.

File: cpp/test/common/sort.cpp [hunk 6]
```diff
@@ -26,12 +27,23 @@ TEMPLATE_TEST_CASE("Test radix sort", "[vector][template]", std::int32_t,
...
-  dolfinx::radix_sort(std::span(vec));
+  dolfinx::radix_sort(vec);
 
   // Check if vector is sorted
   REQUIRE(std::ranges::is_sorted(vec));
 }
 
+TEST_CASE("Test radix sort (projection)", "[radix]")
+{
+  std::vector<int> vec = {3, 6, 2, 1, 5, 4, 0};
```
Comment (jhale) [inline cpp/test/common/sort.cpp:38]: Switch to a modern int type with specific lengths (and perhaps parametrise over two or three types?).

File: cpp/dolfinx/common/sort.h [hunk 7]
```diff
@@ -6,161 +6,105 @@
...
+      for (T c : current_perm)
+        counter[(proj(c) & mask) >> mask_offset]++;
+
+      // Prefix sum to get the inserting position
+      offset[0] = 0;
+      std::partial_sum(counter.begin(), counter.end(),
+                       std::next(offset.begin()));
+      for (T c : current_perm)
+      {
+        std::int32_t bucket = (proj(c) & mask) >> mask_offset;
```
Comment (jhale) [inline cpp/dolfinx/common/sort.h:88]: Correct?

File: cpp/dolfinx/common/sort.h [hunk 8]
```diff
@@ -6,161 +6,105 @@
...
+        std::int32_t new_pos = offset[bucket + 1] - counter[bucket];
```
Comment (jhale) [inline cpp/dolfinx/common/sort.h:89]: Correct?

File: cpp/dolfinx/common/sort.h (hunk 7 above)
Comment (schnellerhase) [reply cpp/dolfinx/common/sort.h]: No :)

**Comment by IgorBaratta** [comment]
Have you tested the radix algorithm for performance regression after these changes? 
It's used in performance-critical parts where we would otherwise use std::sort.

**Comment by jhale** [comment]
Any comments @garth-wells?

**Comment by schnellerhase** [comment]
No haven't done any such thing, how can I run/trigger them properly?

**Comment by jhale** [comment]
@IgorBaratta had some benchmark code for his original PR.

https://github.com/FEniCS/dolfinx/pull/1622

**Comment by schnellerhase** [comment]
Is the code that produced these plots still around?

**Comment by IgorBaratta** [comment]
> Is the code that produced these plots still around?

The first test is just the creation of a cube with increasing number of cells.
You can use `dolfinx::list_timings` to query timing of create entities.

**Comment by jhale** [comment]
Don't worry about boost:: and std:: etc. this was discussed before - just test against the radix in `main`.

**Comment by schnellerhase** [comment]
What compile flags do I need to pass for a benchmark build? I'm assuming `CMAKE_BUILD_TYPE=Release` and what for the pip build?

**Comment by jhale** [comment]
> What compile flags do I need to pass for a benchmark build? I'm assuming `CMAKE_BUILD_TYPE=Release` and what for the pip build?

I'd recommend doing the benchmark purely from C++. Release will give you -O3 which is fine.

File: cpp/dolfinx/common/sort.h [hunk 9]
```diff
@@ -6,161 +6,105 @@
...
-  std::span<T> next_perm = buffer;
-  for (int i = 0; i < its; i++)
+  /// Sort a range with radix sorting algorithm. The bucket
+  /// size is determined by the number of bits to sort at a time (2^BITS).
+  /// @tparam R Type of range to be sorted.
+  /// @tparam P Projection type to be applied on range elements to produce a
+  /// sorting index.
+  /// @tparam BITS The number of bits to sort at a time.
+  /// @param[in, out] range The range to sort.
+  /// @param[in] P Element projection.
```
Comment (garth-wells) [inline cpp/dolfinx/common/sort.h:34]: Could some more explanation be added? Would be nice to support with an example.

File: cpp/dolfinx/common/sort.h [hunk 10]
```diff
@@ -6,161 +6,105 @@
...
-
-  std::vector<std::int32_t> perm2(perm.size());
-  std::span<std::int32_t> current_perm = perm;
-  std::span<std::int32_t> next_perm = perm2;
-  for (int i = 0; i < its; i++)
-  {
-    // Zero counter
-    std::ranges::fill(counter, 0);
+    // Sort N bits at a time
+    constexpr int bucket_size = 1 << BITS;
```
Comment (garth-wells) [inline cpp/dolfinx/common/sort.h:53]: Is `int` the appropriate type?

File: cpp/dolfinx/common/sort.h [hunk 11]
```diff
@@ -6,161 +6,105 @@
...
+    T mask = (T(1) << BITS) - 1;
 
-    // Count number of elements per bucket
-    for (auto cp : current_perm)
+    // Compute number of iterations, most significant digit (N bits) of
+    // maxvalue
+    int its = 0;
```
Comment (garth-wells) [inline cpp/dolfinx/common/sort.h:58]: Is the the appropriate integer type?

File: cpp/dolfinx/common/sort.h [hunk 12]
```diff
@@ -6,161 +6,105 @@
...
-      T value = array[cp] - *min;
-      std::int32_t bucket = (value & mask) >> mask_offset;
-      std::int32_t pos = offset[bucket + 1] - counter[bucket];
-      next_perm[pos] = cp;
-      counter[bucket]--;
+      // Zero counter array
+      std::ranges::fill(counter, 0);
+
+      // Count number of elements per bucket
+      for (const auto& c : current_perm)
```
Comment (garth-wells) [inline cpp/dolfinx/common/sort.h:79]: Could this just be a copy rather than a reference?

File: cpp/dolfinx/common/sort.h [hunk 13]
```diff
@@ -6,161 +6,105 @@
...
+        counter[(proj(c) & mask) >> mask_offset]++;
+
+      // Prefix sum to get the inserting position
+      offset[0] = 0;
+      std::partial_sum(counter.begin(), counter.end(),
+                       std::next(offset.begin()));
+      for (const auto& c : current_perm)
```
Comment (garth-wells) [inline cpp/dolfinx/common/sort.h:86]: See above comment on reference vs copy.

File: cpp/dolfinx/common/sort.h [hunk 14]
```diff
@@ -189,7 +133,8 @@ std::vector<std::int32_t> sort_by_perm(std::span<const T> x, std::size_t shape1)
     int col = shape1 - 1 - i;
     for (std::size_t j = 0; j < shape0; ++j)
       column[j] = x[j * shape1 + col];
-    argsort_radix<T, BITS>(column, perm);
+
+    radix_sort(perm, [&](auto index) { return column[index]; });
```
Comment (garth-wells) [inline cpp/dolfinx/common/sort.h:137]:
```suggestion
    radix_sort(perm, [&column](auto index) { return column[index]; });
```

File: cpp/dolfinx/fem/DirichletBC.cpp [hunk 15]
```diff
@@ -358,7 +358,8 @@ std::array<std::vector<std::int32_t>, 2> fem::locate_dofs_topological(
   // Remove duplicates
   std::vector<std::int32_t> perm(bc_dofs[0].size());
   std::iota(perm.begin(), perm.end(), 0);
-  dolfinx::argsort_radix<std::int32_t>(bc_dofs[0], perm);
+  dolfinx::radix_sort(perm, [&](auto index) { return bc_dofs[0][index]; });
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.cpp:361]:
```suggestion
  dolfinx::radix_sort(perm, [&dofs = bc_dofs[0]](auto index) { return dofs[index]; });
```

File: cpp/dolfinx/fem/DirichletBC.cpp [hunk 16]
```diff
@@ -413,7 +414,9 @@ std::array<std::vector<std::int32_t>, 2> fem::locate_dofs_topological(
     // Remove duplicates and sort
     perm.resize(sorted_bc_dofs[0].size());
     std::iota(perm.begin(), perm.end(), 0);
-    dolfinx::argsort_radix<std::int32_t>(sorted_bc_dofs[0], perm);
+    dolfinx::radix_sort(perm,
+                        [&](auto index) { return sorted_bc_dofs[0][index]; });
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.cpp:418]:
```suggestion
                        [&dofs = sorted_bc_dofs[0]](auto index) { return dofs[index]; });
```

File: cpp/dolfinx/refinement/utils.cpp [hunk 17]
```diff
@@ -257,7 +257,8 @@ std::array<std::vector<std::int32_t>, 2> refinement::transfer_facet_meshtag(
   // Sort values into order, based on facet indices
   std::vector<std::int32_t> sort_order(tag_values.size());
   std::iota(sort_order.begin(), sort_order.end(), 0);
-  dolfinx::argsort_radix<std::int32_t>(facet_indices, sort_order);
+  dolfinx::radix_sort(sort_order,
+                      [&](auto index) { return facet_indices[index]; });
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/utils.cpp:261]:
```suggestion
                      [&facet_indices](auto index) { return facet_indices[index]; });
```

File: cpp/dolfinx/refinement/utils.cpp [hunk 18]
```diff
@@ -351,7 +352,8 @@ refinement::transfer_cell_meshtag(const mesh::MeshTags<std::int32_t>& tags0,
   // Sort values into order, based on cell indices
   std::vector<std::int32_t> sort_order(tag_values.size());
   std::iota(sort_order.begin(), sort_order.end(), 0);
-  dolfinx::argsort_radix<std::int32_t>(cell_indices, sort_order);
+  dolfinx::radix_sort(sort_order,
+                      [&](auto index) { return cell_indices[index]; });
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/utils.cpp:356]:
```suggestion
                      [&cell_indices](auto index) { return cell_indices[index]; });
```

**Review by garth-wells — CHANGES_REQUESTED** [review]
Looks nice - hard to check in detail but test should detect any issues.

Would be good to verify performance.

File: cpp/dolfinx/common/sort.h (hunk 12 above)
Comment (schnellerhase) [reply cpp/dolfinx/common/sort.h:79]: Before these changes it was a copy. Since the value type `T` however now is not restricted to trivially copyable types any longer, we should stick to `const auto&` I think.

**Comment by schnellerhase** [comment]
Results of [benchmark](https://github.com/schnellerhase/dolfinx/pull/24)

on `main`
```bash
2024-07-23T12:00:43+02:00
Running ./sort
Run on (16 X 4768 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 512 KiB (x8)
  L3 Unified 16384 KiB (x1)
Load Average: 3.53, 2.55, 1.80
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
BM_create_box/10/iterations:5        11.9 ms         9.99 ms            5
BM_create_box/50/iterations:5        1159 ms         1159 ms            5
BM_create_box/100/iterations:5      10438 ms        10437 ms            5
```

and on `sort_unique`
```bash
2024-07-23T12:04:02+02:00
Running ./sort
Run on (16 X 4768 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 512 KiB (x8)
  L3 Unified 16384 KiB (x1)
Load Average: 2.21, 2.22, 1.81
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
BM_create_box/10/iterations:5        5.87 ms         5.85 ms            5
BM_create_box/50/iterations:5        1110 ms         1110 ms            5
BM_create_box/100/iterations:5      10147 ms        10147 ms            5
```

**Comment by schnellerhase** [comment]
So it seems that we even achieved quite a nice speed up throughout. But maybe somebody with access to a more stable server for benchmarking could maybe rerun this with a couple large values of `n` as well? This was only run on an average laptop so far, might not show the full picture.

**Comment by jhale** [comment]
This is good enough evidence for me @schnellerhase - we're just looking for a large difference.

File: cpp/dolfinx/common/sort.h (hunk 10 above)
Comment (schnellerhase) [reply cpp/dolfinx/common/sort.h]: Implemented that. No more explicit int usage present.
